Wink Generic Thermostat Support

woopelderly opened the pull request:
<!-- Thanks for contributing to the openHAB project! -->
<!-- Please describe the goal and effect of your PR here. -->
<!-- Pay attention to the below notes and to *the guidelines* for this repository. -->
<!-- Feel free to delete any comment lines in the template (starting with "<!--"). -->

<!-- TITLE -->
[wink][WIP] Initial contribution
Add support for generic thermostats using the existing structure of the wink binding.

<!-- Please provide a PR summary in the *Title* above, according to the following schema: -->
<!-- - If related to one specific addon: Mention the addon shortname in square brackets -->
<!--   e.g. "[exec]", "[netatmo]" or "[tesla]" -->
<!-- - If the PR is work in progress: Add "[WIP]" -->
<!-- - Give a short meaningful description in imperative mood -->
<!--   e.g. "Add support for device XYZ" or "Fix wrongly handled exception" -->
<!--   for a new add-on/binding: "Initial contribution" -->
<!-- Examples: -->
<!-- - "[homematic] Improve communication with weak signal devices" -->
<!-- - "[timemachine][WIP] Initial contribution" -->
<!-- - "Update contribution guidelines on new signing rules" -->

<!-- DESCRIPTION -->
This PR is for the addition of the generic thermostat support to the wink binding.  Currently supports setting the thermostat temperature and setting the thermostate mode of operation, i.e. heat_only, cool_only, or auto.  Many other channels exist for viewing the state of the thermostat and heating/cooling system.  My motivation for adding this is that I have a Emerson Sensi thermostat which is not currently supported by OpenHab, but is supported by Wink.  Therefore, I thought it made sense to add this feature to the Wink binding.  I updated the Readme documentation as well.  The following table shows all the existing channels with short descriptions.:
<table>
<tr><td><b>Thing</b></td><td><b>Channel</b></td><td><b>Description</b></td></tr>
<tr><td rowspan="15">thermostat</td><td>thermostat_currentsetpoint</td><td>Thermostat Current Desired Temperature</td></tr>
<tr><td>thermostat_currentmode</td><td>The current set mode: Cool, Heat, or Auto</td></tr>
<tr><td>thermostat_hasfan</td><td>Whether or not the thermostat unit has a fan, Readonly</td></tr>
<tr><td>thermostat_online</td><td>Whether or not the device is reachable remotely, Readonly</td></tr>
<tr><td>thermostat_ecomode</td><td>Whether or not the thermostat is running in an energy efficient mode, Readonly</td></tr>
<tr><td>thermostat_humidity</td><td>Thermostat Current Humidity, Readonly</td></tr>
<tr><td>thermostat_occupied</td><td>Whether or not the thermostat has detected occupancy in the last 30 minutes, Readonly</td></tr>
<tr><td>thermostat_techname</td><td>Contractor contact data, Readonly</td></tr>
<tr><td>thermostat_techphone</td><td>Contractor contact data, Readonly</td></tr>
<tr><td>thermostat_lasterror</td><td>The current alert/warning on the thermostat, Readonly</td></tr>
<tr><td>thermostat_fanactive</td><td>Whether or not the fan is actively running, Readonly</td></tr>
<tr><td>thermostat_runningmode</td><td>The current running mode: Aux, Cool, Heat, or Idle, Readonly</td></tr>
<tr><td>thermostat_smarttemperature</td><td>Ecobee only, mean temp of all remote sensors and thermostat, Readonly</td></tr>
<tr><td>thermostat_currenttemperature</td><td>Thermostat Current Temperature, maps to room temperature last read from device itself, Readonly</td></tr>
<tr><td>thermostat_externaltemperature</td><td>The outdoor temperature/weather, Readonly</td></tr>
</table>

<!-- Please give a few sentences describing the overall goals of the pull request. -->
<!-- Give enough details to make the improvement and changes of the PR understandable  -->
<!-- to both developers and tech-savy users. -->

<!-- Please keep the following in mind: -->
<!-- - What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ? -->
<!-- - Did you describe the PRs motivation and goal? -->
<!-- - Did you provide a link to any prior discussion, e.g. an issue or community forum thread? -->
<!-- - Did you describe new features for the end user? -->
<!-- - Did you describe any noteworthy changes in usage for the end user? --> 
<!-- - Was the documentation updated accordingly, e.g. the addon README? -->
<!-- - Does your contribution follow the coding guidelines: -->
<!--   http://docs.openhab.org/developers/development/guidelines.html -->
<!-- - Did you check for any (relevant) issues from the static code analysis: -->
<!--   http://docs.openhab.org/developers/development/bindings.html#static-code-analysis -->
<!-- - Did you sign-off your work: -->
<!--   http://docs.openhab.org/developers/contributing/contributing#sign-your-work -->
Signed-off-by: Joel Shuman <joel.shumann@gmail.com> (github: woopelderly)

woopelderly committed:
Added initial support for displaying generic thermostat channels and
setting the desired temperature. - Joel Shuman

woopelderly committed:
Added the ability to change the mode of the thermostat, rather than just
display the mode.

Also, added thermostat information to the wink readme.md.  I moved all
of the PaperUI thing and channel xml config info to its own file.

woopelderly committed:
wink Thermostat state update bug fixes. -Joel Shuman

shawncrosby commented:
I think this looks pretty good.  I'm in a bit of a busy time right now, so I can't really give this a good going over at the moment.  Hopefully next week.  It looks like we're using different white space or something because files you've updated appear to be full re-writes in the compare screen.  I wonder if I have something set incorrectly.  Not sure how thermostat.xml gets read...is it included in thing-types.xml somehow?

shawncrosby commented:
I have a number of changes to make based on the original PR too...I think yours are good though.  I'll let you know if I see something there too.

woopelderly commented:
It is ok if you cannot get to it right away.  Thanks for your initial feedback.  As far as the whitespace, my eclipse IDE used spaces instead of tabs for indentation.  I can switch that and recommit.

woopelderly commented:
Regarding the thermostat.XML, at the top of the file the name of the binding must be there and then all the files must get pulled in.  I did not dig into the framework to see how it did this, but I noticed other bindings had more than one XML file for things and channels.

woopelderly commented:
Do you know how to get the built jar file from this pull request similar to Kai's documentation here?


> Binary Builds from Pull Requests
> Instead of manually building the jar from sources and uploading it to Github/Dropbox/GDrive or some other website for sharing, there is now a much simpler way: I have setup the PR builder on our Jenkins, which first of all validates that a PR is buildable, but secondly also deploys the result in a new Maven repo on our Artifactory instance: https://openhab.jfrog.io/openhab/libs-pullrequest-local/
> 
> So as an example, the Plugwise binding from this PR is automatically available as a jar at https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.plugwise/2.1.0-SNAPSHOT/org.openhab.binding.plugwise-2.1.0-SNAPSHOT.jar. This url can be directly used as the download url for the IoT marketplace.

From this thread:  [https://community.openhab.org/t/distributing-bindings-through-the-iot-marketplace/24491](url)

When I try this for the wink binding I get the jar file that does not contain any of the thermostat code.

woopelderly committed:
Change indentation from spaces to tabs.

woopelderly commented:
It looks like white space can be ignored by adding ?w=1, or &w=1 if other Params are already in the url. 

https://stackoverflow.com/questions/37007300/how-to-ignore-whitespace-in-github-when-comparing

woopelderly commented:
Wondering if you still would be willing to look this over.

woopelderly committed:
Fixed thermostat support.

shawncrosby commented:
Hey sorry I've been mia on this.  I'll take a look soon.

woopelderly commented:
No worries. Let me know if you have any questions.


Joel

On Jul 15, 2018, at 6:50 AM, Shawn Crosby <notifications@github.com<mailto:notifications@github.com>> wrote:


Hey sorry I've been mia on this. I'll take a look soon.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/shawncrosby/openhab2-addons/pull/1#issuecomment-405082790>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AP2Wz8m1q57ojM_LcAewBTET2TCbq0Phks5uGx6IgaJpZM4RurZO>.

shawncrosby commented on addons/binding/org.openhab.binding.wink/src/main/java/org/openhab/binding/wink/client/IWinkDevice.java:
```diff
@@ -1,77 +1,84 @@
+     *
+     * @return An map of state parameters and values for the device parsed from slightly more complex Json
+     */
+    public Map<String, String> getCurrentStateComplexJson();
```
This shouldn't be in the interface.  The interface should never presume the underlying implementation such as json.  I think the better way to do this might be to extend the JsonWinkDevice class and override the getCurrentState method for your particular needs.

woopelderly replied:
I'm working on a better way to parse so I do not need my own method.  thanks for the feedback.

shawncrosby commented on addons/binding/org.openhab.binding.wink/src/main/java/org/openhab/binding/wink/handler/BinarySwitchHandler.java:
```diff
@@ -46,6 +46,11 @@ public void handleWinkCommand(ChannelUID channelUID, Command command) {
         }
     }
+	
+	@Override
```
this should likely be implemented in the BaseWinkDeviceHandler class rather than copy and pasted to each of the subclasses...overridden where necessary.

shawncrosby commented on addons/binding/org.openhab.binding.wink/src/main/java/org/openhab/binding/wink/handler/WinkHub2BridgeHandler.java:
```diff
@@ -1,154 +1,215 @@
+		return client.getDevice(deviceType, uuid);
+	}
+
+	/**
```
This class was meant to be more generic for proxying commands to the hub.  The logic bits should likely be in the thermostat handler and then once the appropriate temperature was determined, then a simple setProperty would do what you want to do.  You can imagine how ugly this class would get after 100 devices if all the logic was in here.

woopelderly replied:
That makes sense.  I have moved the thermostat related methods.

shawncrosby commented on addons/binding/org.openhab.binding.wink/src/main/java/org/openhab/binding/wink/handler/WinkHub2BridgeHandler.java:
```diff
@@ -1,154 +1,215 @@
+	 * @param device The device to change temperature
+	 * @param mode The desired mode of operation, heat_only, cool_only, auto
+	 */
+	public void setThermostatMode(IWinkDevice device, String mode) {
```
Same for this method.  Add the specifics of the device to the device handler and then use this class to just set the values on the hub.

woopelderly replied:
That makes sense.  I have moved the thermostat related methods.

shawncrosby commented on addons/binding/org.openhab.binding.wink/src/main/java/org/openhab/binding/wink/handler/ThermostatHandler.java:
```diff
@@ -0,0 +1,237 @@
+		return jsonData.get("connection").equals("true");
+	}
+
+	private void setDesiredTemperature(float temperature) {
```
Move all the thermostat specific logic from the bridge handler to these methods.

shawncrosby commented on addons/binding/org.openhab.binding.wink/ESH-INF/thing/thermostat.xml:
```diff
@@ -0,0 +1,159 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I have to go back to my pull request to remember what it was exactly I was supposed to do, but I think that the maintainers are asking me to move all this xml stuff into annotations on the handler classes.  I'll take a look when I get a chance and update you.

woopelderly replied:
I looked at the pull request and I think the xml stuff he was referring to was the files in the OSGI-INF directory and not the xml files used by Paper UI for config.  I could be wrong though.

woopelderly committed:
- IWinkDevice: Removed getCurrentStateComplexJson() from the IWinkDevice interface.
- JsonWinkDevice: Modified toMap to allow Gson JsonArray objects to be converted to string types.
- WinkAuthenticationService: Found a TODO item, possibly needs a factory to replace the DummyService.
- AllHandler Classes: Removed connectionStatus abstract method implementation.
- ThermostatHandler: Relocated implementation details from WinkHub2BridgeHandler to the ThermostatHandler.
- ThermostatHandler: Removed 'jsonDataDesired' calls because pubnub updates do not have any json data in the 'desired_data' field.
- WinkBaseThingHandler: Added and improved logging.

Signed-off-by: Joel Shuman joel.shumann@gmail.com (github: woopelderly)

woopelderly commented:
Hello Shawn,
I have committed fixes for everything, except the xml item you mentioned.  I'll try to track down what you are referring to.
Thanks,
Joel

woopelderly committed:
- Many white-space fixes and comment fixes to satisfy the format checker.
- ThermostatHandler:  Fixed bugs related to setting the mode and desired temperature.  Fixed a bug with Celsius/Fahrenheit.  Removed excess debug logging.
- WinkBaseThingHandler:  Added a loop to wait for the auth service to complete configuration before attempting to complete the initialize() method.  Also added some debug logging.

Signed-off-by: Joel Shuman joel.shumann@gmail.com (github: woopelderly)

woopelderly commented:
I have committed everything I have been working on.  I committed fixes for whitespace issues and comment issues brought up by the maven compiler as well.  The ThermostatHandler had a few bugs that I fixed as well.  Please take another look when you get a change and let me know what you think.

Thanks for your help with this,
Joel
